### PR TITLE
[hail/ptypes/PNDArray] remove elementAddress

### PR DESCRIPTION
**Pull request by akotlar**
John, to gain some familiarity with your codebase, a proposed change. Makes the interface 1 method smaller, and reduces a bit of complexity in element value loading. Also fixes a potential source of errors long term: element loading should depend on the representation (as this controls memory layout), and not the elementType passed to the PNDArray constructor.

This came up as I was writing down the invariants for PNDArray for the PTypes design doc. Feel free to push back on this if you have plans for getElementAddress (although if that's the case we should get rid of loadElementToIRIntermediate)

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -2260,10 +2258,8 @@ private class Emit(
 
         new NDArrayEmitter(mb, nDims, shapeArray,
           xP.shape.pType, xP.elementType, setup, ndt.setup, ndt.m) {
-          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
-            val elementLocation = xP.getElementAddress(idxVars, ndAddress, mb)
-            Region.loadIRIntermediate(outputElementPType)(elementLocation)
-          }
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] =
+            xP.loadElementToIRIntermediate(idxVars, ndAddress, mb)
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala]: This may not work, not sure yet why outputElementPType is different from data.pType

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: Oh I see; this is fine, and also a somewhat confusing way to access xP.elementType. Do we need the xP.elementType on NDArrayEmitter?

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/Emit.scala]: I use the `outputElementPType` field in the `emitLoops` method of `NDArrayEmitter`. That's how I know the type of the NDArray that gets build in the end. You could argue that a more reasonable thing to do would be to just have the `NDArrayEmitter` store the type of the NDArray you'd like to get out in the end, especially if we get to a world where there are many NDArray PTypes.

**Comment by johnc1231** [comment]
@akotlar Needs a bump

**Pull request merged**